components/script/dom/notification.rs

```diff
@@ -0,0 +1,348 @@
+
+        Ok(Self {
+            eventtarget: EventTarget::new_inherited(),
+            title,
+            body,
+            data,
+            dir,
+            image,
+            icon,
+            badge,
+            lang,
+            silent,
+            origin,
+            timestamp,
+            renotify,
+            tag,
+            require_interaction,
+            actions,
+            closed: Cell::new(false),
+        })
+    }
+}
+
+impl NotificationMethods<crate::DomTypeHolder> for Notification {
+    /// <https://notifications.spec.whatwg.org/#constructors>
+    fn Constructor(
+        global: &GlobalScope,
+        proto: Option<HandleObject>,
+        can_gc: CanGc,
+        title: DOMString,
+        options: RootedTraceableBox<NotificationOptions>,
+    ) -> Fallible<DomRoot<Notification>> {
+        // step 1: Check global is a ServiceWorkerGlobalScope
+        if global.downcast::<ServiceWorkerGlobalScope>().is_some() {
+            return Err(Error::Type("Can not call in a service worker.".to_string()));
+        }
+
+        // step 2: Check options.actions must be empty
+        if !options.actions.is_empty() {
+            return Err(Error::Type("Actions must be empty.".to_string()));
+        }
+
+        // step 3: Create a notification with a settings object
+        // https://notifications.spec.whatwg.org/#create-a-notification-with-a-settings-object
+        let fallback_timestamp = global.performance().Now().to_u64().unwrap();
+        let notification = Notification::new(
+            global,
+            proto,
+            can_gc,
+            title,
+            options,
+            None,
+            None,
+            fallback_timestamp,
+        )?;
+
+        // step 5.1: Check permission
+        let state = get_descriptor_permission_state(PermissionName::Notifications, Some(global));
+        if state == PermissionState::Granted {
+            // TODO: step 5.2: Run permission show step
+            todo!()
+        } else {
+            notification
+                .upcast::<EventTarget>()
+                .fire_event(atom!("error"), CanGc::note());
+        }
+
+        Ok(notification)
+    }
+
+    /// <https://notifications.spec.whatwg.org/#get-the-notifications-permission-state>
+    fn GetPermission(global: &GlobalScope) -> Fallible<NotificationPermission> {
+        let state = get_descriptor_permission_state(PermissionName::Notifications, Some(global));
+        match state {
+            PermissionState::Granted => Ok(NotificationPermission::Granted),
+            PermissionState::Denied => Ok(NotificationPermission::Denied),
+            PermissionState::Prompt => Ok(NotificationPermission::Default),
+        }
+    }
+
+    /// <https://notifications.spec.whatwg.org/#dom-notification-requestpermission>
+    fn RequestPermission(
+        _global: &GlobalScope,
+        _permission_callback: Option<Rc<NotificationPermissionCallback>>,
+    ) -> Rc<Promise> {
+        todo!()
```

Comment:
Ditto.

Reply:
Since this method needs a return value, I'll give it a dummy `Promise` for now and leave a // TODO comment